### PR TITLE
feat: Tag component

### PR DESCRIPTION
**Pull request by sarahmonster**
This is a draft; there are a few things I think we'll need to add once we have the docs ironed out, but I'm unclear on a few things. 

Here's the draft doc: https://www.notion.so/octopusthink/Tag-1bf7e10623104926a6560ba1b1b43d06

Once that's stable, I'll move them into our README here and see about adding other props as needed. Here's what I think we'll need (marked my name next to ones I should be able to handle pretty easily.)

- [x] Ability to add a `title` prop or visually hidden text. 
- [x] A grouping of semantic status Tags (probably with themeable colours!) (@sarahmonster)
- [x] A prop for numerical Tags + tweaked styling. (@sarahmonster)
- [x] Theme variables for Tags. (@sarahmonster)
- [x] OnClose events for dismissable Tags.
- [x] OnClick events and/or href navigation for interactive Tags. (Probably using the Link component underneath?)
- [x] Probably needs some more tests.
- [x] Should extract the colour contrast into its own utilities file so we can use it elsewhere as needed. (I'd love feedback on how best to offer up error messages in the case that a colour just doesn't have sufficient contrast though!) (@sarahmonster)

### CODE REVIEW

**Comment by sarahmonster** [comment]
Quick question: right now, we ask for two props to make a Tag dismissable:

> If a Tag is dismissable, it can be passed an `onDismiss` event that will be called with the tag's "dismiss" button is activated.

`<Tag color="hotpink" dismissable onDismiss={onDismiss}>Hot pink</Tag>`

Seems like it would make more sense to have just the one—if you pass an `onDismiss` event, the Tag will automatically inherit the `dismissable` styling. I'm not sure there's a sensible use case where you'd want to use one without the other.... unless I'm not thinking about something in particular?

**Comment by sarahmonster** [comment]
Okay! So everything on my list above is more or less done now, except for the theme variables, but I'm thinking it probably makes sense to implement those at a slightly later phase where we're more focussed on the themeing, since there's going to be a wee bit of complexity to consider there, but if we think it's worth doing now I can look at it tomorrow.

Here's what I need help with:
- Fixing my linter errors! (I can fix the fourth one, but the `import/no-unresolved`s might be an issue with the Link component, and the `react/jsx-one-expression-per-line` is forming a circular loop with Prettier, so I'm not sure how to fix those.)
-  OnClick events and/or href navigation for interactive Tags. (Probably using the Link component underneath?)
- Probably needs some more tests.

🙌

File: src/components/ui/Tag/index.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            font-style: bold;
+            position: absolute;
+            left: -10000px;
+            top: auto;
+            width: 1px;
+            height: 1px;
+            overflow: hidden;
+          `}
+        >
+          {label}:
```
Comment (tofumatt) [inline src/components/ui/Tag/index.js]: So, this is failing the linter for a kind of dumb reason: Prettier and the linter are disagreeing with each other. That's fine and fixable (it's rare but it happens, and the underlying rules are usually useful).

But this is problematic for an unrelated reason: it's super-locale specific. Other languages may not use colons here. This should be a config option or something, but we aren't there yet with locale strings, so I think for now this should _just_ be `{label}`. If that means we need to use an `aria-describedby` prop we can, though I don't think it's needed. But I'll remove this colon for now.

File: src/components/ui/Tag/index.js [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+import { Icon } from 'components/ui/Icon';
+import { getContrastingTextColor, metadata, toUnits } from 'styles';
+import { useTheme } from 'themes';
+
+export const Tag = forwardRef((props, ref) => {
+  const theme = useTheme();
+  const { children, label, onDismiss, ...otherProps } = props;
+
+  return (
+    <span ref={ref} {...otherProps}>
```
Comment (tofumatt) [inline src/components/ui/Tag/index.js]: This is why you're getting the `count` error; you're passing `otherProps`, which includes `otherProps.count` in here to a `<span>`, which doesn't support the `count` prop. You need to "extract" it above:

```jsx
  const { children, count, label, onDismiss, ...otherProps } = props;
```

File: src/components/ui/Tag/README.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+Nautilus defaults to showing tag labels in small-caps to distinguish them from body copy and interactive elements.
+
+Small caps can be difficult for users to parse, so it's important to use them sparingly. Remember that Tags are intended only to present *supplementary* data that isn't critical to understanding the core content. Tags should *never* contain content or data that's essential to understanding the content you're presenting.
+
+Use sentence case when passing text strings to a Tag. Small caps styling will be added by CSS. If you pass a word in all caps, screen readers can behave unpredictably. **Only** pass labels in all caps if they're abbreviations and should be read out loud as such.
+
+### Grouping Tags
+
+When showing a group or list of tags, use a Group or List component to indicate that tags are part of the same grouping.
```
Comment (tofumatt) [inline src/components/ui/Tag/README.md]: Do we plan on providing a `Group` component? It seems pretty low-level and I don't like that kind of "developer-knowledge-required" accessibility approach. I'd rather we mandate a `<Tags.Tag>`/`<Tags.Item>` be a child of `<Tags>` and do it automatically for the output. If a user supplies only a single tag we could modify the markup accordingly and omit the grouping—and we could be smart with an `aria-describedby` element instead of prefixing the label on every tag. But this approach requires the developers work harder or rememeber when to group tags. I don't understand when multiple tags like this *wouldn't* be related. Or if they are, I'd rather *that* be the explicit option someone has to choose.

I feel like we can be better-by-default here.

File: src/components/ui/Tag/index.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+};
+
+Tag.propTypes = {
+  /** @ignore */
+  children: PropTypes.node,
+  /** Use colour to differentiate different tags or indicate status. */
+  color: PropTypes.string,
+  /** Optional label for providing additional context to screen reader users. */
+  label: PropTypes.string,
+  /** Style numbers in a more badge-y kind of way. */
```
Comment (tofumatt) [inline src/components/ui/Tag/index.js]: We can tweak this explanation I think—it's a very cute Sarah-ism, but might read strangely to an ESL user 😄

File: src/components/ui/Tag/index.js [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  count: PropTypes.bool,
```
Comment (tofumatt) [inline src/components/ui/Tag/index.js]: Count is the wrong prop name I think. Why not `badge`?

File: src/themes/nautilus/index.js [hunk 6]
```diff
@@ -211,6 +220,15 @@ export const theme = {
           strokeWidth: 1.125,
         },
       },
+
+      // TODO: implement these!
+      Tag: {
```
Comment (tofumatt) [inline src/themes/nautilus/index.js]: We should probably use a "z" here. And these should be added to the theme prop-types. Maybe I'll organise some tests to verify the prop-types are up-to-date...

Comment (sarahmonster) [reply src/themes/nautilus/index.js]: ....Tagz?

Comment (tofumatt) [reply src/themes/nautilus/index.js]: I have no idea where that came from? Oh, I know! Copy/pasted from the wrong line.

File: src/themes/nautilus/index.js [hunk 7]
```diff
@@ -211,6 +220,15 @@ export const theme = {
...
+        capitalisation: 'uppercase', // uppercase, lowercase, sentence, title
```
Comment (tofumatt) [inline src/themes/nautilus/index.js]: This should use a "z".

File: src/themes/nautilus/index.js [hunk 8]
```diff
@@ -211,6 +220,15 @@ export const theme = {
...
+        colourPosition: 'background', // background or border
```
Comment (tofumatt) [inline src/themes/nautilus/index.js]: Use American spelling for props.

File: src/components/ui/Tag/index.js (hunk 1 above)
Comment (sarahmonster) [reply src/components/ui/Tag/index.js]: But then to screen readers, it'll read "Category books" or possibly even "Categorybooks", which seems sub-optimal. Is there a better way we can approach this?

File: src/themes/nautilus/index.js [hunk 9]
```diff
@@ -211,6 +220,15 @@ export const theme = {
...
+        borderRadius: 0,
+        padding:
+          'toUnits(theme.spacing.padding.xSmall) toUnits(theme.spacing.padding.small)',
```
Comment (tofumatt) [inline src/themes/nautilus/index.js]: This is a string, not string interpolation, so I have no idea how this is working 😆

Comment (tofumatt) [reply src/themes/nautilus/index.js]: I'm thinking you want:
```suggestion
          `${toUnits(theme.spacing.padding.xSmall)} ${toUnits(theme.spacing.padding.small)}`,
```

File: src/components/ui/Tag/index.js (hunk 1 above)
Comment (tofumatt) [reply src/components/ui/Tag/index.js]: Using a different element and `aria-describedby` might be best. This approach is bad for localisation and will require awkward config. Described by might handle that for us; it's worth checking out to see if that's true.

File: src/components/ui/Tag/index.js (hunk 2 above)
Comment (sarahmonster) [reply src/components/ui/Tag/index.js]: Ohhhh I didn't think I needed to do that, since `count` isn't being used by this function. Does that mean I need to pull in `status` as well?

Comment (tofumatt) [reply src/components/ui/Tag/index.js]: Correct! It's a bit awkward because, as you say: it isn't being used. But yeah, it's how those object spreads work. 🤷‍♂

File: src/components/ui/Tag/README.md (hunk 3 above)
Comment (sarahmonster) [reply src/components/ui/Tag/README.md]: I left responses in Notion but maybe the notifications aren't going through. Let me copy here!

> I'm not sure that makes sense in all cases though—what if we just want a single Tag, for a date or a status maybe?

> Maybe we should have an OPTIONAL <TagGroup> or <Tags> container?

Basically. SOME Tags will warrant a group, but others will not. We need to allow for some way of using both. <Tags.Tag> feels really awkward and fragile when you only need a single Tag. Maybe it would make sense instead to have an optional TagGroup or TagList component?

File: src/components/ui/Tag/index.js (hunk 5 above)
Comment (sarahmonster) [reply src/components/ui/Tag/index.js]: I suppose we could do that. My concern is that this then starts bleeding into the styling, whereas the intention here is to use it for counts or tallies or numbers. I'm trying to aim for more semantic, rather than stylistic, names, but badge is commonly enough used by other systems that I'm okay with going more relaxed about it for now.

File: src/components/ui/Tag/README.md (hunk 3 above)
Comment (tofumatt) [reply src/components/ui/Tag/README.md]: Oops, nope, they are. I just haven't looked. I'll review there.

File: src/themes/nautilus/index.js (hunk 9 above)
Comment (sarahmonster) [reply src/themes/nautilus/index.js]: It's not! It's just a placeholder. Ultimately I'd like for it to reference the spacing values declared in the theme, but right now we don't have a good way of doing that until I refactor some theme variables, which is part of why I don't want to deal with implementing theme variables just yet.

File: src/components/ui/Tag/index.js (hunk 5 above)
Comment (tofumatt) [reply src/components/ui/Tag/index.js]: But it's a boolean. If this is supposed to take a number wouldn't the _content_ be a number? This is just about styling from what I understand.

File: src/themes/nautilus/index.js (hunk 6 above)
Comment (sarahmonster) [reply src/themes/nautilus/index.js]: (Where are the theme prop-types stored?)

File: src/components/ui/Tag/index.js (hunk 5 above)
Comment (sarahmonster) [reply src/components/ui/Tag/index.js]: No no, it's not supposed to take a number—it's just a property of "this is a count". (It was originally called `numerical` but I changed it out for being too wordy.) The idea here is that "badge" only refers to how the styling is applied, but "count" or "number" refers to actual semantic usage—like "sidebar" rather than "blue-box" in CSS classes, for instance.

I could see how that might be confusing though. Alternative suggestions I'm fine with using here:
- badge
- number
- numeric
- tally

File: src/themes/nautilus/index.js (hunk 6 above)
Comment (tofumatt) [reply src/themes/nautilus/index.js]: https://github.com/octopusthink/nautilus/blob/master/src/themes/propTypes.js

But don't sweat it for now. I just mentioned it for awareness/posterity really.

File: src/components/ui/Tag/index.js (hunk 2 above)
Comment (sarahmonster) [reply src/components/ui/Tag/index.js]: Why is it that `status` doesn't generate an error then? (Or `colour`, for that matter, or other props used similarly in other components?) Is this something we need to be checking for?

File: src/components/ui/Tag/index.js (hunk 5 above)
Comment (tofumatt) [reply src/components/ui/Tag/index.js]: Yeah, count and tally are definitely misleading and I find them confusing. I think `badge` is the right approach. What semantic value is there in saying the kind of content here? And why only offer this one option? I think it's rooted in the right place but `badge` is the right approach here.

`isNumeric` or `isNumber` could work otherwise but I don't love that. I'd say stick with `badge`. 🤷‍♂

File: src/components/ui/Tag/index.js (hunk 2 above)
Comment (tofumatt) [reply src/components/ui/Tag/index.js]: Because they're `undefined` by default and ES6 spreads omit keys with a value of `undefined` through. But because this is `false` by default it gets passed through.

We're already checking for it and it's working. 😄

File: src/themes/nautilus/index.js (hunk 6 above)
Comment (sarahmonster) [reply src/themes/nautilus/index.js]: I was definitely not aware that file even existed. 😮 

Probably worth doing a release focussed on themeing in the near-ish future (maybe 0.6 or 0.7) and looking into these sorts of things in a bit more depth—there are a number of things I think we can improve about the structure of the themeing itself, and I think they're probably worth tackling once we have Foundation and Function a bit more fleshed out.

File: src/themes/nautilus/index.js (hunk 9 above)
Comment (sarahmonster) [reply src/themes/nautilus/index.js]: (I'm leaving this as is because toUnits isn't available here and it's really just there as a placeholder for when we actually implement this.)

**Comment by sarahmonster** [comment]
Per discussion, let's implement the following changes here:

- [ ] require all Tags to be wrapped in a Tag component
- [ ] include a label prop on the Tags component for titling component

```
<Tags label="Date">
    <Tags.Tag>12 September 2019</Tags.Tag>
  </Tags>
```

- [ ] onClick events or whatever for tags that want to be interactive

There's some potentially-helpful information to peruse/steal/borrow from here: https://a11y-guidelines.orange.com/web_EN/exemples/tag/index.html

@tofumatt, when you've a moment, could you look into the above for me? I'm not exactly sure how best to handle those changes myself, which means I'm probably wandering out of my depth now.
